Garrison wall game

Hazado opened the pull request:
Imports these changes from moul-scripts for the Wall Game in Gahreesen
https://github.com/TOC-Shard/moul-scripts/commit/b92f3e7e952cac0128663acb829e06f9aa35190c#diff-1820401ce4e8b6ceaaefa00e66a7e94537c5be337dff74865468cee48b8b3310
https://github.com/TOC-Shard/Plasma/commit/5e5bf3dd48b3df55a90fb229cdab81e28e8f0d76

Edits made by me to get it compiled.

Hoikas commented on Sources/Plasma/PubUtilLib/plAvatar/plAvBrainHuman.cpp:
```diff
@@ -472,14 +474,23 @@ bool plAvBrainHuman::MsgReceive(plMessage * msg)
-        plKey seekTarget = msg->fTarget;
-        plAvTaskSeek *seekTask = new plAvTaskSeek(seekTarget);
-        QueueTask(seekTask);
+        // Warp the player to the Seekpoint
+        plSceneObject *avatarObj = plSceneObject::ConvertNoRef(plNetClientMgr::GetInstance()->GetLocalPlayer());
+        plSceneObject *obj = plSceneObject::ConvertNoRef(msg->fTarget->ObjectIsLoaded());
+        plArmatureMod *localAvatar = plAvatarMgr::GetInstance()->GetLocalAvatar();
+        plArmatureMod *climbAvatar = plArmatureMod::ConvertNoRef(fArmature);
+        if (climbAvatar == localAvatar) // is it our avatar who has to seek?
+        {
+            hsMatrix44 target = obj->GetLocalToWorld();
+            plWarpMsg *warp = new plWarpMsg(nullptr, avatarObj->GetKey(), plWarpMsg::kFlushTransform, target);
+            warp->SetBCastFlag(plMessage::kNetPropagate);
+            plgDispatch::MsgSend(warp);
+        }
```
suspicious

Hoikas commented on Sources/Plasma/PubUtilLib/plAvatar/plAvBrainHuman.cpp:
```diff
@@ -81,6 +81,8 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 #include "plMessage/plLOSRequestMsg.h"
 #include "plMessage/plRideAnimatedPhysMsg.h"
 #include "plMessage/plSwimMsg.h"
+#include "pnMessage/plWarpMsg.h"
```
fixme

Hoikas replied:
Please alphabetize this include with the rest of the nucleus level includes.

Hoikas commented on Scripts/SDL/Garrison.sdl:
```diff
@@ -912,3 +912,81 @@ STATEDESC Garrison
+	VAR BOOL	grsnDnElevSwitchOn[1]		DEFAULT=0 DEFAULTOPTION=VAULT	
+	VAR BOOL	grsnDoorSwitchOn[1]		DEFAULT=0 DEFAULTOPTION=VAULT	
+	VAR BOOL	grsnGearBrake01On[1]		DEFAULT=1 DEFAULTOPTION=VAULT	
+	VAR BOOL	grsnGearBrake02On[1]		DEFAULT=1 DEFAULTOPTION=VAULT
+
+# prison variables
+
+	VAR BOOL grsnPrisonBones01vis[1] DEFAULT=0 DEFAULTOPTION=VAULT 
+    	VAR BOOL grsnPrisonBones02vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonBowls01vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonBowls02vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonBowls03vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonBowls04vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonBowls05vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonChains01vis[1] DEFAULT=0 DEFAULTOPTION=VAULT 
+    	VAR BOOL grsnPrisonChains02vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonDirt01vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonDirt02vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonMattress01vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonMattress02vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonWindow01vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    	VAR BOOL grsnPrisonWindow02vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+	VAR BOOL grsnYeeshaPage02Vis[1] DEFAULT=0 DEFAULTOPTION=VAULT
+
+
+# climbing wall variables
+
+	VAR INT			nChairOccupant[1] DEFAULT=-1 DEFAULTOPTION=VAULT
+	VAR INT			sChairOccupant[1] DEFAULT=-1 DEFAULTOPTION=VAULT
+	VAR INT			nWallPlayer[1] DEFAULT=-1 DEFAULTOPTION=VAULT
+	VAR INT			sWallPlayer[1] DEFAULT=-1 DEFAULTOPTION=VAULT
+	VAR BYTE		grsnClimbingWallEnabled[1] DEFAULT=1 DEFAULTOPTION=VAULT
+    VAR BYTE        nState[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    VAR BYTE        sState[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    VAR BYTE        NumBlockers[1] DEFAULT=0 DEFAULTOPTION=VAULT
+    VAR BYTE        nBlockerChange[2] DEFAULT=0 #1-index; 2-on
+    VAR BYTE        sBlockerChange[2] DEFAULT=0
+    VAR INT         northWall[20] DEFAULT=-1 #Blockers set for each team
+    VAR INT         southWall[20] DEFAULT=-1
+}
```
cleanup whitespace

Hoikas commented on Scripts/Python/grsnWallImagerDisplayS.py:
```diff
@@ -43,95 +43,46 @@
+        #We only get a notify from nState
+        value = self.ageSDL[VARname][0]
+        if(value == kWait):
+            for blocker in self.ageSDL["southWall"]:
+                if(blocker == -1):
+                    break
+                pImager.value[blocker].runAttachedResponder(kBlockerOn)
```
misleading

Hoikas commented on Scripts/Python/grsnWallImagerDisplayN.py:
```diff
@@ -43,97 +43,50 @@
+        #We only get a notify from nState
+        value = self.ageSDL[VARname][0]
+        if(value == kWait):
+            for blocker in self.ageSDL["northWall"]:
+                if(blocker == -1):
+                    break
+                yImager.value[blocker].runAttachedResponder(kBlockerOn)
```
misleading

Hoikas commented on Scripts/Python/grsnWallImagerDisplayN.py:
```diff
@@ -43,97 +43,50 @@
-                    i = i + 1
-"""
+        if(len(PtGetPlayerList()) and self.ageSDL["nState"] >= kWait):
+            for blocker in self.ageSDL["northWall"]:
+                if(blocker == -1):
+                    return
+                yImager.value[blocker].runAttachedResponder(kBlockerOn)
```
misleading

Hoikas commented on Scripts/Python/grsnWallEventHandler.py:
```diff
@@ -0,0 +1,172 @@
+    def HandlePQuit(self, idx):
+        stage = 'PQuit' + str(idx)
+        msg = kWallPrefix + PtGetLocalizedString("Gahreesen.Wall.PQuit" + str(idx))
+        return (stage, msg)
```
EOF

Hoikas commented on Scripts/Python/grsnWallEventHandler.py:
```diff
@@ -0,0 +1,172 @@
+                self.startTime = PtGetDniTime()
+            if(value == kEnd):
+                sec = PtGetDniTime() - self.startTime
+                msg = kWallPrefix + PtGetLocalizedString("Gahreesen.Wall.MatchDuration", [time.strftime("%M:%S", time.localtime(sec))])
```
new strings

Hoikas commented on Scripts/Python/grsnWallEventHandler.py:
```diff
@@ -0,0 +1,172 @@
+
+#constants
+kBlockerTimer = 0
+kWallPrefix = "TOC Wall: "
```
TOC specific

Hoikas commented on Scripts/Python/grsnWallConstants.py:
```diff
@@ -0,0 +1,86 @@
+kEventYQuit = "HandleYQuit"
+kEventPQuit = "HandlePQuit"
+
+eventHandler = None
+def InitEventHandler(instance):
+    global eventHandler
+    eventHandler = instance
```
suspicious, dependent on internal implementation of PFMs

Hoikas commented on Scripts/Python/grsnMainWallPython.py:
```diff
@@ -43,222 +43,76 @@
-        PtDebugPrint("grsnMainWallPython::init end")        
-"""    
+        self.version = 2
+        self.ageSDL = None
```
should probably not cache this

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+pPanelSound = ptAttribResponder(58,"purple panel sound",['main','up','down','select','blockerOn','blockerOff','gameStart','denied'],netForce=0)
 
+#TOC specific
+panelLightAnimP = ptAttribMaterialAnimation(61, "Purple Panel Light Animation")
+panelLightAnimY = ptAttribMaterialAnimation(62, "Yellow Panel Light Animation")
+sitPanelLightsY = ptAttribSceneobjectList(63, "Yellow sitting indicators")
+sitPanelLightsP = ptAttribSceneobjectList(64, "Purple sitting indicators")
+confirmPanelLightsY = ptAttribSceneobjectList(65, "Yellow ready indicators")
+confirmPanelLightsP = ptAttribSceneobjectList(66, "Purple ready indicators")
```
validate these and change everything back to north/south to maintain consistency with named python files (or rename those in the PFMs)

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+confirmPanelLightsY = ptAttribSceneobjectList(65, "Yellow ready indicators")
+confirmPanelLightsP = ptAttribSceneobjectList(66, "Purple ready indicators")
 
 ##############################################################
 # grsnWallPython
 ##############################################################
```
document the difference between grsnWallPython (control panels) vs grsnMainWallPython (the game board)

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
-                PtDebugPrint("tubes open")
+        self.version = 5
+        self.BlockerCount = 0
+        self.ageSDL = None
```
should probably not cache this

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+            if(self.ageSDL["nState"][0] == kSit):
+                self.ResetWall(resetState=False)
+                self.ChangeGameState(kYellow, kSelectCount)
+                yPanelSound.run(self.key, state='main')
```
respond to sdl

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+            ### Confirm Blocker ###
+            elif(self.ageSDL["nState"][0] == kSetBlocker):
+                if(self.GetNumBlockerSet(kYellow) < self.ageSDL["NumBlockers"][0]):
+                    yPanelSound.run(self.key, state='denied')
```
respond to sdl

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+                    yPanelSound.run(self.key, state='denied')
+                else:
+                    self.ChangeGameState(kYellow, kWait)
+                    yTubeOpen.run(self.key)
+                    yPanelSound.run(self.key, state='gameStart')
```
respond to sdl

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+                    yTubeOpen.run(self.key)
+                    yPanelSound.run(self.key, state='gameStart')
+                    if(PtFindAvatar(events) == PtGetLocalAvatar()):
+                        self.ageSDL["nChairOccupant"] = (cID,)
```
who cares?

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+            if(state == kSelectCount):
+                for i in range(0,20):
+                    if(i < self.ageSDL["NumBlockers"][0]):
+                        exec (team + "CountLights.value[i].runAttachedResponder(kRedFlash)")
+                    elif((team == kYellow and i < self.oldCounty) or (team == kPurple and i < self.oldCountp)):
+                        exec (team + "CountLights.value[i].runAttachedResponder(kRedOff)")
+                exec('self.oldCount' + team + ' = self.ageSDL["NumBlockers"][0]')
+            elif(state >= kSetBlocker):
+                for i in range(0,self.ageSDL["NumBlockers"][0]):
+                    if(i < self.GetNumBlockerSet(team)):
+                        exec (team + "CountLights.value[i].runAttachedResponder(kTeamLightsOn)")
```
!!!! security hole

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
-                        #self.SetWallIndex(index,False,False)
-                        self.ChangeBlockerState(False,index,False)
-"""
+                        exec (team + "CountLights.value[i].runAttachedResponder(kRedOn)")
```
!!!! security hole

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
-                        self.ChangeBlockerState(False,index,False)
-"""
+                        exec (team + "CountLights.value[i].runAttachedResponder(kRedOn)")
+            return
```
```suggestion
```

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+            exec ("tenBtn" + team + ".disable()")
+            exec ("fifteenBtn" + team + ".disable()")
+        elif(state == kSelectCount):
+            for i in range(0,171):
+                self.SetPanelBlocker(team, i, False, onlyLight=True)
+                exec (team + "Panel.value[i].physics.disable()")
+            exec ("goBtn" + team + "Object.value.runAttachedResponder(kDim)")
+            exec ("goButton" + team + ".disable()")
+            exec ("upButton" + team + ".enable()")
+            exec ("dnButton" + team + ".enable()")
+            exec ("readyButton" + team + ".enable()")
+            exec ("fiveBtn" + team + ".enable()")
+            exec ("tenBtn" + team + ".enable()")
+            exec ("fifteenBtn" + team + ".enable()")
+            for i in range(0,20):
+                if(i < self.ageSDL["NumBlockers"][0]):
+                    exec (team + "CountLights.value[i].runAttachedResponder(kRedFlash)")
+                else:
+                    exec (team + "CountLights.value[i].runAttachedResponder(kRedOff)")
+        elif(state == kSetBlocker):
+            exec ("goBtn" + team + "Object.value.runAttachedResponder(kPulse)")
+            exec ("goButton" + team + ".enable()")
+            exec ("upButton" + team + ".disable()")
+            exec ("dnButton" + team + ".disable()")
+            exec ("readyButton" + team + ".disable()")
+            exec ("fiveBtn" + team + ".disable()")
+            exec ("tenBtn" + team + ".disable()")
+            exec ("fifteenBtn" + team + ".disable()")
+            for i in range(0,self.ageSDL["NumBlockers"][0]):
+                if(i < self.GetNumBlockerSet(team)):
+                    exec (team + "CountLights.value[i].runAttachedResponder(kTeamLightsOn)")
+                else:
+                    exec (team + "CountLights.value[i].runAttachedResponder(kRedOn)")
+            for i in range(0,171):
+                exec (team + "Panel.value[i].physics.enable()")
+        else:
+            exec ("goBtn" + team + "Object.value.runAttachedResponder(kDim)")
+            exec ("goButton" + team + ".disable()")
+            for i in range(0,171):
+                exec (team + "Panel.value[i].physics.disable()")
```
!!!! security hole

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+            elif(team == kPurple):
+                self.ageSDL["sState"] = (state,)
+
+    def TOCPanelLight(self):
+        yState = self.ageSDL["nState"][0]
+        pState = self.ageSDL["sState"][0]
+
+        panelLightAnimP.animation.speed(0.25)
+        panelLightAnimP.animation.stop()
+        panelLightAnimP.animation.skipToTime(0.16)
+        panelLightAnimY.animation.speed(0.25)
+        panelLightAnimY.animation.stop()
+        panelLightAnimY.animation.skipToTime(0.16)
+
+        if(yState == kSetBlocker):
+            panelLightAnimY.animation.play()
+        if(pState == kSetBlocker):
+            panelLightAnimP.animation.play()
+
+        for i in range(0,2):
+            if(yState == kSit or yState == kSelectCount):
+                sitPanelLightsY.value[i].draw.enable()
+                confirmPanelLightsY.value[i].draw.disable()
+            elif(yState == kWait or yState == kEntry or yState == kGameInProgress):
+                sitPanelLightsY.value[i].draw.enable()
+                confirmPanelLightsY.value[i].draw.enable()
+            elif(yState == kStandby or yState == kEnd):
+                sitPanelLightsY.value[i].draw.disable()
+                confirmPanelLightsY.value[i].draw.disable()
+
+            if(pState == kSit or pState == kSelectCount):
+                sitPanelLightsP.value[i].draw.enable()
+                confirmPanelLightsP.value[i].draw.disable()
+            elif(pState == kWait or pState == kEntry or pState == kGameInProgress):
+                sitPanelLightsP.value[i].draw.enable()
+                confirmPanelLightsP.value[i].draw.enable()
+            elif(pState == kStandby or pState == kEnd):
+                sitPanelLightsP.value[i].draw.disable()
+                confirmPanelLightsP.value[i].draw.disable()
```
fiddly

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,773 @@
+        self.TOCPanelLight() #TOC
+
+    def SetPanelBlocker(self,team,id,on,onlyLight=False):
+        if(onlyLight):
+            if(on):
+                exec (team + "Lights.value[id].runAttachedResponder(kTeamLightsOn)")
+            else:
+                exec (team + "Lights.value[id].runAttachedResponder(kTeamLightsOff)")
+            return
+
+        if(team == kYellow):
+            sdl = "northWall"
+        else:
+            sdl = "southWall"
+
+        if(on):
+            idx = self.GetNumBlockerSet(team) #push it to the end
+            if(idx >= self.ageSDL["NumBlockers"][0]): #all Blockers set
+                exec (team + "PanelSound.run(self.key, state='denied')")
+            else:
+                if(self.IAmMaster()):
+                    self.ageSDL.setIndex(sdl,idx,id)
+                exec (team + "PanelSound.run(self.key, state='blockerOn')")
+                self.SetPanelBlocker(team, id, True, onlyLight=True)
+        else:
+            wall = list(self.ageSDL[sdl])
+            idx = 0
+            for blocker in wall:
+                if(blocker == id):
+                    break
+                idx += 1
+            if(idx >= 20):
+                PtDebugPrint("grsnWallPython::SetPanelBlocker: Blocker %d not found in %s" % (id, sdl))
+            else:
+                for i in range(idx,19): #rearrange the array, so all the blockers come at the beginning
+                    wall[i] = wall[i+1]
+                wall[19] = -1
+                if(self.IAmMaster()):
+                    self.ageSDL[sdl] = tuple(wall)
+                exec (team + "PanelSound.run(self.key, state='blockerOff')")
```
!!!! security hole

Hoikas reviewed (commented):
These are notes for myself, you don't have to act on these @Hazado

TODO
- Rewrite fiddly single case/duplicated team code to generic code instantiated for each team
- Fix security holes
- Ensure actions happen in response to synch changes, not local notifications
- Remove condition that the panel user is the only one who can climb
- Validate plAvTaskClimb's local player manipulations
- Make consistent: north/south, black/white, purple/yellow, whatever (note those pairing orders may not be correct)

Hazado commented:
> These are notes for myself, you don't have to act on these @Hazado
> 
> TODO
> * Make consistent: north/south, black/white, purple/yellow, whatever (note those pairing orders may not be correct)

changed it to north/south

Yellow is North and uses White room nexus on game end
Purple is South and uses Black room nexus on game end

Hazado commented:
Issues ive noticed and thoughts

Singleplayer wall, needs a check for entering both chutes...currently will warp you to both if you do it
Nexus needs to be de-subworlded and elevator code added
Rewrite of python code for security reasons and general sanity
Allow multiple climbers for both sides
Beeping sound at chair panels probably needs to be turned off when the game starts
Possible reward for doing game, is maintainer suit...not sure what the maintainer suit being generally available would break

EhrenCG commented:
> Yellow is North and uses White room nexus on game end
> Purple is South and uses Black room nexus on game end

Did you try using both the buttons to forfeit and climbing to the top of the walls?  I think the forfeit buttons might have the sides mixed up.

> Beeping sound at chair panels probably needs to be turned off when the game starts

Yes, I believe the beeping only continues if you enter in the number of blockers on the yellow control panel, and the beeping stops if you enter the blockers on the purple one.  Also, notice that you can add/remove blockers on the yellow panel while standing up, but not on the purple one.

Hazado commented:
Removed the "exec" code, which should fix the security issues.

Hazado commented:
Think I have got a working version for multiple climbers per side
Would appreciate another review of the changes I have made in the last few days

Hazado commented:
> Did you try using both the buttons to forfeit and climbing to the top of the walls? I think the forfeit buttons might have the sides mixed up.

Yes, i've used both forfeit buttons and climbed to the top of the walls.
If the forfeit buttons take you to the wrong sides, ill be able to fix that. 
Will be sure to test that out once we start doing a multiplayer test, will be much easier to figure that out once I have a second person with me to see if the winner and forfeiters go to the same nexus.

Hazado commented:
I tried making the nexus elevator into a jalak style instead of a subworld...it had issues.
When it would get to the top of the elevator, the avatar would warp to the new floor.
	I tried putting a hole in the floor in the exact shape as the elevator floor, still warped.
	Set the second floor mesh to be non-static, still warped.
When you ran on the elevator floor, you would randomly be moved up like you walked on something higher than the floor.

Hazado commented:
Brought this up-to-date with the current minkata testing.

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,1181 @@
+        # create shallow copy of full edge array
+        myEdges = list(edgeList)
+
+        for i in range(0,20):
+            blockerId = ageSDL[team][i]
+            if blockerId == -1:
```
```suggestion
        for blockerId in ageSDL[team]:
            if blockerId == -1:
```

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -44,1203 +44,1181 @@
+                # this blocker has no effect on edges, it's a throwaway
+                pass
+
+        adjacencyList = dict()
+        for nodeA, nodeB in myEdges:
+            adjacencyList.setdefault(nodeA,[])
+            adjacencyList.setdefault(nodeB,[])
```
```suggestion
        adjacencyList = defaultdict(list)
        for nodeA, nodeB in myEdges:
```

Add
```python
from collections import defaultdict
```
to the top of the file with the rest of the imports.

Hoikas commented on Scripts/Python/ki/__init__.py:
```diff
@@ -3748,7 +3748,11 @@ def BigKISetStatics(self):
-        ageText.setStringW(ageName)
+        #Gahreesen Maintainer Nexii use private chat channels 6 and 7
+        if self.chatMgr.privateChatChannel == 6 or self.chatMgr.privateChatChannel == 7:
+            ageText.setStringW("Maintainer Nexus")
```
Can we get this in a loc file or somewhere it can be localized?

Hoikas commented on Scripts/Python/ki/__init__.py:
```diff
@@ -3748,7 +3748,11 @@ def BigKISetStatics(self):
         ageName = GetAgeName().replace("(null)", "").strip()
         PtDebugPrint("xKI.BigKISetStatics(): Displaying age name of {}.".format(ageName), level=kDebugDumpLevel)
-        ageText.setStringW(ageName)
+        #Gahreesen Maintainer Nexii use private chat channels 6 and 7
+        if self.chatMgr.privateChatChannel == 6 or self.chatMgr.privateChatChannel == 7:
```
Can we use constants instead of magic numbers?

Hoikas commented on Sources/Plasma/PubUtilLib/plAvatar/plAnimStage.cpp:
```diff
@@ -300,7 +300,7 @@ bool plAnimStage::ISendNotify(uint32_t notifyMask, uint32_t notifyType, plArmatu
         msg->AddMultiStageEvent(stageNum, notifyType, armature->GetTarget(0)->GetKey());
 
-        if (!genBrain || !genBrain->RelayNotifyMsg(msg))
+        if (stageNum < 0 || !genBrain || !genBrain->RelayNotifyMsg(msg))
```
What is the goal of this change? I'm a little concerned that this may have unintended side effects because, IIRC, stage `-1` means something.

Hoikas commented on Sources/Plasma/PubUtilLib/plAvatar/plAvBrainHuman.cpp:
```diff
@@ -472,14 +474,23 @@ bool plAvBrainHuman::MsgReceive(plMessage * msg)
     if(isStartClimb)
     {
-        // let's build a seek task to get us to the attach point
-        plKey seekTarget = msg->fTarget;
-        plAvTaskSeek *seekTask = new plAvTaskSeek(seekTarget);
-        QueueTask(seekTask);
+        // Warp the player to the Seekpoint
+        plSceneObject *avatarObj = plSceneObject::ConvertNoRef(plNetClientMgr::GetInstance()->GetLocalPlayer());
+        plSceneObject *obj = plSceneObject::ConvertNoRef(msg->fTarget->ObjectIsLoaded());
+        plArmatureMod *localAvatar = plAvatarMgr::GetInstance()->GetLocalAvatar();
+        plArmatureMod *climbAvatar = plArmatureMod::ConvertNoRef(fArmature);
+        if (climbAvatar == localAvatar) // is it our avatar who has to seek?
+        {
+            hsMatrix44 target = obj->GetLocalToWorld();
+            plWarpMsg *warp = new plWarpMsg(nullptr, avatarObj->GetKey(), plWarpMsg::kFlushTransform, target);
+            warp->SetBCastFlag(plMessage::kNetPropagate);
+            plgDispatch::MsgSend(warp);
+        }
```
Did the seeking not work?

Hazado replied:
current seeking can randomly crash, or place the avatar further down than the start point when starting to climb on the wall.

Hoikas commented:
Can we please revert all of the whitespace changes to Garrison.sdl and instead only include the new version of the `STATEDESC`?

Hoikas commented:
Can we get more detail in the commit message of https://github.com/H-uru/Plasma/pull/912/commits/9f72c65cc5cc424010fdcaa63f4f8f7e8ac140e5 to indicate the desired fixes?

ZarothYe commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -40,1207 +40,1187 @@
+                            for item in wornItem:
+                                DefaultColor1.append(avatar.avatar.getTintClothingItem(item[0],1))
+                                DefaultColor2.append(avatar.avatar.getTintClothingItem(item[0],2))
+                        i = 1
+                        for item in clothing[0:]:
+                            avatar.avatar.netForce(1)
+                            avatar.avatar.wearClothingItem(item, 0)
+                            avatar.avatar.tintClothingItem(item, DefaultColor1[i], 0)
+                            avatar.avatar.tintClothingItemLayer(item, DefaultColor2[i], 2, 1)
+                            i += 1
```
What's the reasoning for starting the index at 1 instead of 0 here? Is it possible to rework this area and the South tube multibehavior block below to use `zip` instead of the `i`-indexed loop just like you do in a couple other places like https://github.com/H-uru/Plasma/blob/9f72c65cc5cc424010fdcaa63f4f8f7e8ac140e5/Scripts/Python/grsnWallPython.py#L306-L310

Hazado replied:
1 is used because the wornitem, which we got the colors from initially has the face in the 0 position.
1 is the first item we are wearing.
This is why the clothing was rearranged in https://github.com/H-uru/Plasma/commit/9f72c65cc5cc424010fdcaa63f4f8f7e8ac140e5 , without doing that you would wear the suit and put the wrong colors the wrong clothing slots.

ZarothYe commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -40,1207 +40,1187 @@
+                            for item in wornItem:
+                                DefaultColor1.append(avatar.avatar.getTintClothingItem(item[0],1))
+                                DefaultColor2.append(avatar.avatar.getTintClothingItem(item[0],2))
+                        i = 1
+                        for item in clothing[0:]:
+                            avatar.avatar.netForce(1)
+                            avatar.avatar.wearClothingItem(item, 0)
+                            avatar.avatar.tintClothingItem(item, DefaultColor1[i], 0)
+                            avatar.avatar.tintClothingItemLayer(item, DefaultColor2[i], 2, 1)
+                            i += 1
```
Same questions as above for the North tube multibehavior block.

ZarothYe commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -40,1207 +40,1187 @@
+                    for item in clothing[0:]:
+                        if (not self.IItemInCloset(avatar, item)):
+                            PtDebugPrint(f'DEBUG: grsnWallPython.OnNotify():  Adding {item} to your closet and wearing it.')
+                            avatar.avatar.addWardrobeClothingItem(item, ptColor().white(), ptColor().white())
```
Should this be adding to the closet with the `white` color? I noticed you used the `DefaultColor1` and `DefaultColor2` for each piece in other places.

Hazado replied:
Since the suit doesnt have any custom colors, adding to the closet with white is fine.

ZarothYe commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -40,1207 +40,1187 @@
+                    for item in clothing[0:]:
+                        if (not self.IItemInCloset(avatar, item)):
+                            PtDebugPrint(f'DEBUG: grsnWallPython.OnNotify():  Adding {item} to your closet and wearing it.')
+                            avatar.avatar.addWardrobeClothingItem(item, ptColor().white(), ptColor().white())
```
Ditto with NorthTeamWin block above -- unsure if using `white` is intended.

Hazado commented:
> Can we please revert all of the whitespace changes to Garrison.sdl and instead only include the new version of the `STATEDESC`?

Done.

Hoikas commented on Scripts/Python/grsnNexusBookMachine.py:
```diff
@@ -51,8 +51,8 @@
 booknorthInPos = ptAttribActivator(4,"North book in position event")
-booksouthOutResponder = ptAttribResponder(5,"South book out")
-booknorthOutResponder = ptAttribResponder(6,"North book out")
+booksouthOutResponder = ptAttribResponder(5,"South book out",netForce=1)
+booknorthOutResponder = ptAttribResponder(6,"North book out",netForce=1)
```
```suggestion
booksouthOutResponder = ptAttribResponder(5,"South book out",netForce=True)
booknorthOutResponder = ptAttribResponder(6,"North book out",netForce=True)
```

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -137,22 +137,22 @@
-    '03_FRFoot_Suit',
-    '03_FLegs_Suit',
-    '03_FTorso_Suit']
+MaleSuit = {
+    0:'03_MLegs_Suit',
+    1:'03_MTorso_Suit',
+    2:'03_MLHand_Suit',
+    3:'03_MRHand_Suit',
+    5:'03_MHAcc_SuitHelmet',
+    6:'03_MLFoot_Suit',
+    7:'03_MRFoot_Suit'}
+FemaleSuit = {
+    0:'03_FLegs_Suit',
+    1:'03_FTorso_Suit',
+    2:'03_FLHand_Suit',
+    3:'03_FRHand_Suit',
+    5:'03_FHair_SuitHelmet',
+    6:'03_FLFoot_Suit',
+    7:'03_FRFoot_Suit'}
```
Why not make these tuples? The keys are just sequential integers, which you get for free with tuples. You're then able to avoid performance penalties for looking these up multiple times in the later code.

Hazado replied:
they are not sequential, 4 is missing

Hoikas replied:
Use a placeholder, like `None` for index 4 in that case. You can skip falsey values in a loop by doing `for suitItem in filter(None, FemaleSuit):`.

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -137,22 +137,22 @@
+    7:'03_FRFoot_Suit'}
     
 wornItem = []
 DefaultColor1 = []
```
You should probably change these to dicts to match the dict comprehension in OnNotify.

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -571,16 +554,13 @@ def OnNotify(self,state,id,events):
-                                DefaultColor2.append(avatar.avatar.getTintClothingItem(item[0],2))
-                        i = 1
-                        for item in clothing[0:]:
+                            DefaultColor1 = {item[1] : avatar.avatar.getTintClothingItem(item[0],1) for item in wornItem}
+                            DefaultColor2 = {item[1] : avatar.avatar.getTintClothingItem(item[0],2) for item in wornItem}
```
```suggestion
                            DefaultColor1 = {item[1]: avatar.avatar.getTintClothingItem(item[0],1) for item in wornItem}
                            DefaultColor2 = {item[1]: avatar.avatar.getTintClothingItem(item[0],2) for item in wornItem}
```

Minor formatting fix.

Hazado commented:
Requested changes pushed.

Hoikas commented on Scripts/Python/grsnNexusBookMachine.py:
```diff
@@ -116,6 +108,7 @@ def OnServerInitComplete(self):
         if PtGetPlayerList():
             return
         resetResponder.run(self.key,avatar=PtGetLocalAvatar())
+        bookPillarSpinning.run(self.key, netForce=True, netPropagate=True)
```
```suggestion
        bookPillarSpinning.run(self.key)
```

Since you have marked this as `netForce=True` above, these kwargs are redundant.

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -667,9 +667,9 @@ def OnNotify(self,state,id,events):
+                if(team == kNorth and ageSDL["nState"][0] == kSetBlocker):
                     self.SetPanelBlocker(kNorth, index, not self.FindBlocker(kNorth, index))
-                if(team == kSouth and ageSDL["sState"][0] == kSetBlocker and cID == ageSDL["sChairOccupant"][0]):
+                if(team == kSouth and ageSDL["sState"][0] == kSetBlocker):
```
Will this change allow players who are standing behind the occupant to operate the control panel?

Hazado replied:
this was the wrong place to do this
Currently yes anyone that stands inside the region can use the panel buttons

Hoikas commented:
Please rebase this on top of the latest master so the CI failure will resolve.

Hoikas commented:
Is this up to date with what will be provided to MOULa?

Hazado commented:
> Is this up to date with what will be provided to MOULa?
No will update tonight.

Hazado commented:
Updated to match current MOULA release candidate on Staging.

Hoikas commented on Scripts/Python/grsnNexusBookMachine.py:
```diff
@@ -202,6 +228,7 @@ def OnNotify(self,state,id,events):
             PtDebugPrint("grsnNexusBookMachine.OnNotify:\tadding you to private chat channel %d" % (chatChannel.value),level=kDebugDumpLevel)
             PtSendPrivateChatList(self.GetPlayersInChatDistance())
             PtSendKIMessageInt(kSetPrivateChatChannel, chatChannel.value)
+            AreWeInRoom = True
```
This looks like it's vulnerable to problems if someone is running modified Python. I think the initial if statement needs to also include a check against `PtFindAvatar(events) == PtGetLocalAvatar()` to ensure we only fire for our own notifications. The cutscene camera is vulnerable to the same problem.

Hoikas commented on Scripts/Python/grsnWallPython.py:
```diff
@@ -684,6 +684,8 @@ def OnTimer(self,id):
         avatar = PtGetLocalAvatar()
         if(id == kWearOriginalClothes and wornItem):
             for item in wornItem:
+                if item[1] == 8:
```
What does this magic number signify?

Hazado replied:
Accessories, can give it a named variable

Hazado commented on Scripts/Python/grsnNexusBookMachine.py:
```diff
@@ -174,11 +200,11 @@ def OnNotify(self,state,id,events):
         if (id == booknorthInPos.id and master):
             PtDebugPrint("North book aligned")
-            booknorthOutResponder.run(self.key)
+            self.SendNote('NorthBook')
```
Needed so that the nexus book pillar functions correctly in a multiplayer environment. It didnt center correctly when the age owner was syncing the animations to other players. What it does now is have the age owner trigger a notify to all players that the book sequence is happening and the responder plays locally for everyone.

Hoikas replied:
There should be no difference in those two situations...

Hazado replied:
And yet in testing on staging there was... Book coming out slightly offcenter to very offcentered over time

ZarothYe commented on Scripts/Python/grsnNexusBookMachine.py:
```diff
@@ -150,13 +167,22 @@ def OnNotify(self,state,id,events):
+            PtDebugPrint("South book deployed")
+            booksouthOutResponder.run(self.key,netPropagate=False)
+
+        if (id == -1 and events[0][1].find('NorthBook') == 0):
```
Missed removing the extraneous parenthesis at the start in the suggestion... `if id == -1 and ...`

Hazado commented:
Suggested changes pushed.
Rebased to latest master.

Fixed a few bugs found late in testing.
1. Private Chat wont clear properly if you relto out of the Maintainer Nexus area. Now clears on each arrival to Relto Age.
2. Fade In/Out when linking from wall game to Maintainer Nexus sometimes failed for unknown reasons. Leading to screen staying black. Moved it to one script so fade out and in is contained to the script that started it.
3. Reduced the time before the maintainer suit is removed. Less chance for it to be visible to players already in Maintainer Nexus area when new player links in.
4. Maintainer Suit chute area would have the exclusion area up if you linked in during an active game. It lead to the force field effect around the chute entrance, making it hard for the arrival to trigger the entrance sequence.

Hoikas commented on Scripts/Python/Personal.py:
```diff
@@ -111,6 +111,10 @@ def OnFirstUpdate(self):
         # this does not downgrade us if we have a normal KI :)
         PtSendKIMessageInt(kUpgradeKILevel, kMicroKI)
 
+        # Clear any private chat that failed to clear properly
+        PtClearPrivateChatList(PtGetLocalAvatar().getKey())
+        PtSendKIMessageInt(kUnsetPrivateChatChannel, 0)
+
```
It might make more sense to put this in `xKI`'s `OnServerInitComplete()` so it will get cleared out for any links to new Ages (ie via the console).

dpogue replied:
Will this break the case where people go into a neighbourhood egg room, close the door, and then link out? Right now I believe that's treated as an effective way to get long-running private chat with a select group of people

Hoikas replied:
Yeah, it does. I would consider the current behavior a bug, though - if you're not in the know about it, the behavior would probably be an unpleasant surprise (why can't anyone hear me?!?!?!).

Hazado replied:
That is a bug, its supposed to clear itself when you link out.

Hazado replied:
> It might make more sense to put this in `xKI`'s `OnServerInitComplete()` so it will get cleared out for any links to new Ages (ie via the console).

I though xKI only initializes once, but if it this runs every age link it would be a better place for it for sure.

Hoikas replied:
OnServerInitComplete runs every time new Age’s state information comes in :)

Hazado replied:
Moved to xki/__init__.py instead.
